### PR TITLE
Add one api: R.addDefault

### PR DESCRIPTION
**Pull request by adispring**
`addDefault` is used to add one default returned value to function. If function's result is `null`, `undefined` or `NaN`, the given default value will be returned.

Discuss link: https://github.com/ramda/ramda/issues/2111

### CODE REVIEW

**Comment by CrossEye** [comment]
Sorry, I don't remember ever seeing this, although I participated in the discussion leading to it.  I'm reopening, because it seems like a reasonable idea.

**Comment by adispring** [comment]
Ok，If the core team support this pr，I'll resolve the conflict, and put the `addDefault` module to proper directory.

**Comment by CrossEye** [comment]
I think this would be a good addition?  Anyone else?  @ramda/core?

**Comment by semmel** [comment]
**Pro:**
- Currying might come handy:
```javascript
var returnsDefaultPhone = R.addDefault(getReceptionPhone());
var findPhoneByName = returnsDefaultPhone(getPhoneFromName));
var findPhoneByRoomNr = returnsDefaultPhone(getPhoneFromRoomNr));
var call = R.compose(callNumber, findPhoneByName);
call('Max');
```
**Cons:**
- API bloat. (It's just`R.compose(R.defaultTo(value), fn)`. Which should be simple for Ramda users to write`(*)`
- For me the name suggests some kind of addition (numeric or size) which I find misleading. What about naming it `returnDefault` ?

`(*)` - provided they know about `R.defaultTo` which I haven't for all the years using Ramda 😳

**Comment by Bradcomp** [comment]
> provided they know about R.defaultTo which I haven't for all the years using Ramda

😆 

I agree that it's questionable whether this is necessary. I don't know how often I would reach for it. If we move forward, what about something like `withDefault` for the name?

**Comment by CrossEye** [comment]
The only advantage of `addDefeult` as a name is the parallel with `addIndex`.  It's not at all necessary, either the name or the function.  I'm mixed about it.  There is a reasonable argument to be made, but it doesn't seem essential.

**Comment by adispring** [comment]
> **Cons:**
> 
> * API bloat. (It's just`R.compose(R.defaultTo(value), fn)`. Which should be simple for Ramda users to write`(*)`

I have encountered some situations that let me write this:
```js
R.compose(R.defaultTo(value), fn)
// or
const maybeNil = fn(...args)
const result = R.defaultTo(maybeNil)
``` 
again and again，if the `fn` is other functions instead of `path` or `prop`, both of which have `***Or` version. 

`addDefault`/`withDefault` maybe alleviate this situation.

**Comment by CrossEye** [comment]
We never did finish this conversation.  What do you think, @adispring, @semmel, @Bradcomp, @customcommander?

**Comment by customcommander** [comment]
I like decorators too. Not strong feeling for or against this new function. I tend to use the `with` prefix for decorators, e.g. `withDefault(42)(fn)`. Do we have any established naming convention for those?

**Comment by semmel** [comment]
> We never did finish this conversation. What do you think, @adispring, @semmel, @Bradcomp, @customcommander?

It *fits well* with Ramda's notion of an "absent/missing" value. Many Ramda functions (`propOr`, `defaultTo`, ...) interpret `null`, `undefined` and `NaN` as such an "absent/missing" value. This is *consistent* with many JavaScript APIs which return one of those to indicate something is missing. `document.querySelector('impossible-element') // -> null` comes into my mind first.

So I consider this beneficial for wrapping third-party APIs.

However,  @davidchambers pointed out in [his comment](https://github.com/ramda/ramda/issues/2111#issuecomment-294781435), that an "absent/missing" value is much better represented as a `Nothing`. 
Unfortunately the discussion in #2111 got quickly bikeshedded into a discussion about sparse arrays. 

If in *your own code*, you sometimes convey a missing item with `null`, `undefined`,  and sometimes with a `Nothing` you'll make reasoning about your code more difficult. Although using `Maybe`s is not everyones cup of tea I cherish *ramda/ramda-fantasy's* [`Maybe.maybe`](https://github.com/ramda/ramda-fantasy/blob/master/docs/Maybe.md#maybemaybe) which does about the same thing as this PR.

So on one hand Ramda features functions (e.g. `ap`, `promap`, ...) which make few sense without *FantasyLand types*, while on the other hand Ramda encourages working with `null`, `undefined`. Covering such a very wide area of application, it could get hard for Ramda to keep the balancing act.

TLDR: Right now I would not use this function.

**Comment by CrossEye** [comment]
I think there isn't enough strong support to include this.  Closing for now.

**Comment by adispring** [comment]
OK, I think composing `R.defaultTo` and other api that need default value is enough.
